File: src/main/java/com/budgetbuddy/service/SpendingPatternService.java
package com.budgetbuddy.service;

import com.budgetbuddy.model.SpendingPattern;
import com.budgetbuddy.model.Transaction;
import com.budgetbuddy.model.User;
import com.budgetbuddy.repository.SpendingPatternRepository;
import com.budgetbuddy.repository.TransactionRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class SpendingPatternService {

    private static final Logger logger = LoggerFactory.getLogger(SpendingPatternService.class);

    @Autowired
    private TransactionRepository transactionRepository;

    @Autowired
    private SpendingPatternRepository spendingPatternRepository;

    /**
     * Analyze transactions and detect spending patterns for a user.
     * Detects daily, weekly, and monthly routines.
     */
    public List<SpendingPattern> detectPatterns(User user) {
        logger.info("Detecting spending patterns for user: {}", user.getId());
        
        List<Transaction> transactions = transactionRepository.findAll()
            .stream()
            .filter(t -> t.getUser().getId().equals(user.getId()))
            .filter(t -> t.getAmount() != null && t.getAmount() < 0) // Only expenses
            .filter(t -> t.getDate().isAfter(LocalDate.now().minusMonths(6))) // Last 6 months
            .collect(Collectors.toList());

        if (transactions.isEmpty()) {
            logger.info("No transactions found for pattern detection");
            return Collections.emptyList();
        }

        List<SpendingPattern> patterns = new ArrayList<>();

        // Detect daily patterns (e.g., coffee every evening)
        patterns.addAll(detectDailyPatterns(user, transactions));

        // Detect weekly patterns (e.g., weekend food orders)
        patterns.addAll(detectWeeklyPatterns(user, transactions));

        // Detect monthly patterns (e.g., utility bills on 1st of month)
        patterns.addAll(detectMonthlyPatterns(user, transactions));

        // Save patterns
        for (SpendingPattern pattern : patterns) {
            spendingPatternRepository.save(pattern);
        }

        logger.info("Detected {} spending patterns for user {}", patterns.size(), user.getId());
        return patterns;
    }

    /**
     * Detect daily patterns (same merchant/category at similar times)
     */
    private List<SpendingPattern> detectDailyPatterns(User user, List<Transaction> transactions) {
        List<SpendingPattern> patterns = new ArrayList<>();

        // Group by category/subcategory and merchant pattern
        Map<String, List<Transaction>> grouped = transactions.stream()
            .filter(t -> t.getPredictedCategory() != null)
            .collect(Collectors.groupingBy(t -> {
                String key = (t.getPredictedCategory() != null ? t.getPredictedCategory() : "Unknown") + "|" +
                             (t.getPredictedSubcategory() != null ? t.getPredictedSubcategory() : "") + "|" +
                             extractMerchantPattern(t.getNarration());
                return key;
            }));

        for (Map.Entry<String, List<Transaction>> entry : grouped.entrySet()) {
            List<Transaction> txs = entry.getValue();
            if (txs.size() < 10) continue; // Need at least 10 occurrences

            String[] parts = entry.getKey().split("\\|");
            String category = parts[0];
            String subcategory = parts.length > 1 ? parts[1] : null;
            String merchantPattern = parts.length > 2 ? parts[2] : null;

            // Check if it's a daily pattern (occurs most days)
            long uniqueDays = txs.stream()
                .map(Transaction::getDate)
                .distinct()
                .count();

            double avgAmount = txs.stream()
                .mapToDouble(t -> Math.abs(t.getAmount()))
                .average()
                .orElse(0.0);

            // If it occurs on more than 30% of days, it's a daily pattern
            long totalDays = ChronoUnit.DAYS.between(
                txs.stream().map(Transaction::getDate).min(LocalDate::compareTo).orElse(LocalDate.now()),
                LocalDate.now()
            );
            
            if (totalDays > 0 && (uniqueDays * 100.0 / totalDays) > 30) {
                SpendingPattern pattern = new SpendingPattern();
                pattern.setUser(user);
                pattern.setPatternType(SpendingPattern.PatternType.DAILY);
                pattern.setCategory(category);
                pattern.setSubcategory(subcategory);
                pattern.setMerchantPattern(merchantPattern);
                pattern.setAverageAmount(avgAmount);
                pattern.setFrequency((int) (uniqueDays * 30 / Math.max(totalDays, 1))); // Frequency per month
                pattern.setConfidenceScore(Math.min(uniqueDays * 1.0 / Math.max(totalDays / 30, 1), 1.0));
                pattern.setFirstObserved(txs.stream().map(Transaction::getDate).min(LocalDate::compareTo).orElse(LocalDate.now()));
                pattern.setLastObserved(txs.stream().map(Transaction::getDate).max(LocalDate::compareTo).orElse(LocalDate.now()));
                pattern.setIsActive(true);

                patterns.add(pattern);
            }
        }

        return patterns;
    }

    /**
     * Detect weekly patterns (e.g., weekend food orders, Friday coffee)
     */
    private List<SpendingPattern> detectWeeklyPatterns(User user, List<Transaction> transactions) {
        List<SpendingPattern> patterns = new ArrayList<>();

        // Group by day of week and category
        Map<DayOfWeek, Map<String, List<Transaction>>> byDayOfWeek = transactions.stream()
            .filter(t -> t.getPredictedCategory() != null)
            .collect(Collectors.groupingBy(
                t -> t.getDate().getDayOfWeek(),
                Collectors.groupingBy(t -> 
                    (t.getPredictedCategory() != null ? t.getPredictedCategory() : "Unknown") + "|" +
                    (t.getPredictedSubcategory() != null ? t.getPredictedSubcategory() : "") + "|" +
                    extractMerchantPattern(t.getNarration())
                )
            ));

        for (Map.Entry<DayOfWeek, Map<String, List<Transaction>>> dayEntry : byDayOfWeek.entrySet()) {
            DayOfWeek dayOfWeek = dayEntry.getKey();
            
            for (Map.Entry<String, List<Transaction>> categoryEntry : dayEntry.getValue().entrySet()) {
                List<Transaction> txs = categoryEntry.getValue();
                if (txs.size() < 4) continue; // Need at least 4 occurrences

                String[] parts = categoryEntry.getKey().split("\\|");
                String category = parts[0];
                String subcategory = parts.length > 1 ? parts[1] : null;
                String merchantPattern = parts.length > 2 ? parts[2] : null;

                // Check if it occurs consistently on this day
                long weeksWithTransaction = txs.stream()
                    .map(t -> t.getDate().with(DayOfWeek.MONDAY)) // Group by week
                    .distinct()
                    .count();

                long totalWeeks = ChronoUnit.WEEKS.between(
                    txs.stream().map(Transaction::getDate).min(LocalDate::compareTo).orElse(LocalDate.now()),
                    LocalDate.now()
                );

                if (totalWeeks > 0 && (weeksWithTransaction * 100.0 / totalWeeks) > 40) {
                    double avgAmount = txs.stream()
                        .mapToDouble(t -> Math.abs(t.getAmount()))
                        .average()
                        .orElse(0.0);

                    SpendingPattern pattern = new SpendingPattern();
                    pattern.setUser(user);
                    pattern.setPatternType(SpendingPattern.PatternType.WEEKLY);
                    pattern.setCategory(category);
                    pattern.setSubcategory(subcategory);
                    pattern.setMerchantPattern(merchantPattern);
                    pattern.setDayOfWeek(dayOfWeek);
                    pattern.setAverageAmount(avgAmount);
                    pattern.setFrequency((int) weeksWithTransaction);
                    pattern.setConfidenceScore(Math.min(weeksWithTransaction * 1.0 / Math.max(totalWeeks, 1), 1.0));
                    pattern.setFirstObserved(txs.stream().map(Transaction::getDate).min(LocalDate::compareTo).orElse(LocalDate.now()));
                    pattern.setLastObserved(txs.stream().map(Transaction::getDate).max(LocalDate::compareTo).orElse(LocalDate.now()));
                    pattern.setIsActive(true);

                    patterns.add(pattern);
                }
            }
        }

        return patterns;
    }

    /**
     * Detect monthly patterns (e.g., utility bills on 1st, rent on 15th)
     */
    private List<SpendingPattern> detectMonthlyPatterns(User user, List<Transaction> transactions) {
        List<SpendingPattern> patterns = new ArrayList<>();

        // Group by day of month and category
        Map<Integer, Map<String, List<Transaction>>> byDayOfMonth = transactions.stream()
            .filter(t -> t.getPredictedCategory() != null)
            .collect(Collectors.groupingBy(
                t -> t.getDate().getDayOfMonth(),
                Collectors.groupingBy(t -> 
                    (t.getPredictedCategory() != null ? t.getPredictedCategory() : "Unknown") + "|" +
                    (t.getPredictedSubcategory() != null ? t.getPredictedSubcategory() : "") + "|" +
                    extractMerchantPattern(t.getNarration())
                )
            ));

        for (Map.Entry<Integer, Map<String, List<Transaction>>> dayEntry : byDayOfMonth.entrySet()) {
            Integer dayOfMonth = dayEntry.getKey();
            
            for (Map.Entry<String, List<Transaction>> categoryEntry : dayEntry.getValue().entrySet()) {
                List<Transaction> txs = categoryEntry.getValue();
                if (txs.size() < 3) continue; // Need at least 3 occurrences

                String[] parts = categoryEntry.getKey().split("\\|");
                String category = parts[0];
                String subcategory = parts.length > 1 ? parts[1] : null;
                String merchantPattern = parts.length > 2 ? parts[2] : null;

                // Check if it occurs consistently on this day of month
                long monthsWithTransaction = txs.stream()
                    .map(t -> t.getDate().withDayOfMonth(1)) // Group by month
                    .distinct()
                    .count();

                long totalMonths = ChronoUnit.MONTHS.between(
                    txs.stream().map(Transaction::getDate).min(LocalDate::compareTo).orElse(LocalDate.now()),
                    LocalDate.now()
                );

                if (totalMonths > 0 && (monthsWithTransaction * 100.0 / totalMonths) > 50) {
                    double avgAmount = txs.stream()
                        .mapToDouble(t -> Math.abs(t.getAmount()))
                        .average()
                        .orElse(0.0);

                    SpendingPattern pattern = new SpendingPattern();
                    pattern.setUser(user);
                    pattern.setPatternType(SpendingPattern.PatternType.MONTHLY);
                    pattern.setCategory(category);
                    pattern.setSubcategory(subcategory);
                    pattern.setMerchantPattern(merchantPattern);
                    pattern.setDayOfMonth(dayOfMonth);
                    pattern.setAverageAmount(avgAmount);
                    pattern.setFrequency((int) monthsWithTransaction);
                    pattern.setConfidenceScore(Math.min(monthsWithTransaction * 1.0 / Math.max(totalMonths, 1), 1.0));
                    pattern.setFirstObserved(txs.stream().map(Transaction::getDate).min(LocalDate::compareTo).orElse(LocalDate.now()));
                    pattern.setLastObserved(txs.stream().map(Transaction::getDate).max(LocalDate::compareTo).orElse(LocalDate.now()));
                    pattern.setIsActive(true);

                    patterns.add(pattern);
                }
            }
        }

        return patterns;
    }

    /**
     * Extract merchant pattern from narration (simplified merchant name)
     */
    private String extractMerchantPattern(String narration) {
        if (narration == null || narration.isEmpty()) {
            return "UNKNOWN";
        }

        // Extract key words (uppercase, remove UPI codes, etc.)
        String cleaned = narration.toUpperCase()
            .replaceAll("UPI[^\\s]*", "")
            .replaceAll("\\d+", "")
            .replaceAll("[^A-Z\\s]", " ")
            .trim();

        // Take first few significant words
        String[] words = cleaned.split("\\s+");
        if (words.length > 0) {
            return words[0] + (words.length > 1 ? " " + words[1] : "");
        }

        return "UNKNOWN";
    }

    /**
     * Get active patterns for a user
     */
    public List<SpendingPattern> getActivePatterns(User user) {
        return spendingPatternRepository.findByUserAndIsActiveTrue(user);
    }

    /**
     * Refresh patterns for a user (re-analyze and update)
     */
    public void refreshPatterns(User user) {
        // Deactivate old patterns
        List<SpendingPattern> oldPatterns = spendingPatternRepository.findByUser(user);
        for (SpendingPattern pattern : oldPatterns) {
            pattern.setIsActive(false);
            spendingPatternRepository.save(pattern);
        }

        // Detect new patterns
        detectPatterns(user);
    }
}

File: mobile-version/app/src/main/java/com/budgetbuddy/mobile/service/SpendingPatternService.kt
package com.budgetbuddy.mobile.service

import com.budgetbuddy.mobile.data.dao.SpendingPatternDao
import com.budgetbuddy.mobile.data.dao.TransactionDao
import com.budgetbuddy.mobile.data.model.SpendingPattern
import com.budgetbuddy.mobile.data.model.Transaction
import kotlinx.coroutines.flow.first
import java.time.LocalDate
import java.time.temporal.ChronoUnit

class SpendingPatternService(
    private val spendingPatternDao: SpendingPatternDao,
    private val transactionDao: TransactionDao
) {
    
    /**
     * Detect spending patterns for a user
     */
    suspend fun detectPatterns(userId: Long): List<SpendingPattern> {
        val transactions = transactionDao.getTransactionsByUser(userId).first()
            .filter { it.amount < 0 } // Only expenses
            .filter { it.date.isAfter(LocalDate.now().minusMonths(6)) } // Last 6 months
        
        if (transactions.isEmpty()) {
            return emptyList()
        }
        
        val patterns = mutableListOf<SpendingPattern>()
        
        // Detect daily patterns
        patterns.addAll(detectDailyPatterns(userId, transactions))
        
        // Detect weekly patterns
        patterns.addAll(detectWeeklyPatterns(userId, transactions))
        
        // Detect monthly patterns
        patterns.addAll(detectMonthlyPatterns(userId, transactions))
        
        // Save patterns
        patterns.forEach { spendingPatternDao.insertPattern(it) }
        
        return patterns
    }
    
    private fun detectDailyPatterns(userId: Long, transactions: List<Transaction>): List<SpendingPattern> {
        val patterns = mutableListOf<SpendingPattern>()
        
        // Group by category/subcategory
        val grouped = transactions
            .filter { it.predictedCategory != null }
            .groupBy { 
                "${it.predictedCategory}|${it.predictedSubcategory ?: ""}|${extractMerchantPattern(it.narration)}"
            }
        
        grouped.forEach { (key, txs) ->
            if (txs.size < 10) return@forEach // Need at least 10 occurrences
            
            val parts = key.split("|")
            val category = parts[0]
            val subcategory = parts.getOrNull(1)?.takeIf { it.isNotEmpty() }
            val merchantPattern = parts.getOrNull(2)?.takeIf { it.isNotEmpty() }
            
            val uniqueDays = txs.map { it.date }.distinct().count()
            val avgAmount = txs.map { kotlin.math.abs(it.amount) }.average()
            
            val totalDays = ChronoUnit.DAYS.between(
                txs.minOfOrNull { it.date } ?: LocalDate.now(),
                LocalDate.now()
            )
            
            // If it occurs on more than 30% of days, it's a daily pattern
            if (totalDays > 0 && (uniqueDays * 100.0 / totalDays) > 30) {
                patterns.add(
                    SpendingPattern(
                        userId = userId,
                        patternType = SpendingPattern.PatternType.DAILY,
                        category = category,
                        subcategory = subcategory,
                        merchantPattern = merchantPattern,
                        averageAmount = avgAmount,
                        frequency = (uniqueDays * 30 / totalDays.coerceAtLeast(1)).toInt(),
                        confidenceScore = (uniqueDays * 1.0 / (totalDays / 30).coerceAtLeast(1)).coerceAtMost(1.0),
                        firstObserved = txs.minOfOrNull { it.date },
                        lastObserved = txs.maxOfOrNull { it.date },
                        isActive = true,
                        createdAt = LocalDate.now()
                    )
                )
            }
        }
        
        return patterns
    }
    
    private fun detectWeeklyPatterns(userId: Long, transactions: List<Transaction>): List<SpendingPattern> {
        val patterns = mutableListOf<SpendingPattern>()
        
        // Group by day of week and category
        val byDayOfWeek = transactions
            .filter { it.predictedCategory != null }
            .groupBy { it.date.dayOfWeek.value } // Use value (1-7) instead of DayOfWeek enum
            .mapValues { (_, txs) ->
                txs.groupBy { 
                    "${it.predictedCategory}|${it.predictedSubcategory ?: ""}|${extractMerchantPattern(it.narration)}"
                }
            }
        
        byDayOfWeek.forEach { (dayOfWeekValue, categoryGroups) ->
            categoryGroups.forEach { (key, txs) ->
                if (txs.size < 4) return@forEach // Need at least 4 occurrences
                
                val parts = key.split("|")
                val category = parts[0]
                val subcategory = parts.getOrNull(1)?.takeIf { it.isNotEmpty() }
                val merchantPattern = parts.getOrNull(2)?.takeIf { it.isNotEmpty() }
                
                val weeksWithTransaction = txs.map { 
                    it.date.minusDays((it.date.dayOfWeek.value - 1).toLong()) // Get Monday of that week
                }.distinct().count()
                
                val totalWeeks = ChronoUnit.WEEKS.between(
                    txs.minOfOrNull { it.date } ?: LocalDate.now(),
                    LocalDate.now()
                )
                
                if (totalWeeks > 0 && (weeksWithTransaction * 100.0 / totalWeeks) > 40) {
                    val avgAmount = txs.map { kotlin.math.abs(it.amount) }.average()
                    
                    patterns.add(
                        SpendingPattern(
                            userId = userId,
                            patternType = SpendingPattern.PatternType.WEEKLY,
                            category = category,
                            subcategory = subcategory,
                            merchantPattern = merchantPattern,
                            dayOfWeek = dayOfWeekValue,
                            averageAmount = avgAmount,
                            frequency = weeksWithTransaction.toInt(),
                            confidenceScore = (weeksWithTransaction * 1.0 / totalWeeks.coerceAtLeast(1)).coerceAtMost(1.0),
                            firstObserved = txs.minOfOrNull { it.date },
                            lastObserved = txs.maxOfOrNull { it.date },
                            isActive = true,
                            createdAt = LocalDate.now()
                        )
                    )
                }
            }
        }
        
        return patterns
    }
    
    private fun detectMonthlyPatterns(userId: Long, transactions: List<Transaction>): List<SpendingPattern> {
        val patterns = mutableListOf<SpendingPattern>()
        
        // Group by day of month and category
        val byDayOfMonth = transactions
            .filter { it.predictedCategory != null }
            .groupBy { it.date.dayOfMonth }
            .mapValues { (_, txs) ->
                txs.groupBy { 
                    "${it.predictedCategory}|${it.predictedSubcategory ?: ""}|${extractMerchantPattern(it.narration)}"
                }
            }
        
        byDayOfMonth.forEach { (dayOfMonth, categoryGroups) ->
            categoryGroups.forEach { (key, txs) ->
                if (txs.size < 3) return@forEach // Need at least 3 occurrences
                
                val parts = key.split("|")
                val category = parts[0]
                val subcategory = parts.getOrNull(1)?.takeIf { it.isNotEmpty() }
                val merchantPattern = parts.getOrNull(2)?.takeIf { it.isNotEmpty() }
                
                val monthsWithTransaction = txs.map { 
                    it.date.withDayOfMonth(1) 
                }.distinct().count()
                
                val totalMonths = ChronoUnit.MONTHS.between(
                    txs.minOfOrNull { it.date } ?: LocalDate.now(),
                    LocalDate.now()
                )
                
                if (totalMonths > 0 && (monthsWithTransaction * 100.0 / totalMonths) > 50) {
                    val avgAmount = txs.map { kotlin.math.abs(it.amount) }.average()
                    
                    patterns.add(
                        SpendingPattern(
                            userId = userId,
                            patternType = SpendingPattern.PatternType.MONTHLY,
                            category = category,
                            subcategory = subcategory,
                            merchantPattern = merchantPattern,
                            dayOfMonth = dayOfMonth,
                            averageAmount = avgAmount,
                            frequency = monthsWithTransaction.toInt(),
                            confidenceScore = (monthsWithTransaction * 1.0 / totalMonths.coerceAtLeast(1)).coerceAtMost(1.0),
                            firstObserved = txs.minOfOrNull { it.date },
                            lastObserved = txs.maxOfOrNull { it.date },
                            isActive = true,
                            createdAt = LocalDate.now()
                        )
                    )
                }
            }
        }
        
        return patterns
    }
    
    private fun extractMerchantPattern(narration: String?): String {
        if (narration.isNullOrEmpty()) return "UNKNOWN"
        
        val cleaned = narration.uppercase()
            .replace(Regex("UPI[^\\s]*"), "")
            .replace(Regex("\\d+"), "")
            .replace(Regex("[^A-Z\\s]"), " ")
            .trim()
        
        val words = cleaned.split("\\s+".toRegex())
        return if (words.isNotEmpty()) {
            words[0] + if (words.size > 1) " ${words[1]}" else ""
        } else "UNKNOWN"
    }
    
    suspend fun getActivePatterns(userId: Long): List<SpendingPattern> {
        return spendingPatternDao.getActivePatternsByUser(userId).first()
    }
    
    suspend fun refreshPatterns(userId: Long) {
        // Deactivate old patterns
        spendingPatternDao.deactivateAllPatternsForUser(userId)
        
        // Detect new patterns
        detectPatterns(userId)
    }
}

